### PR TITLE
MAGN-5816: Crash when slider using slider in run atuo

### PR DESCRIPTION
**Pull request by pboyer**
### Issue

At the heart of it, this was a race condition.  When `PreviewControl.RefreshExpandedDisplay` was enumerating the items in an `Array` from `MirrorData`, the engine was modifying that data.  This could result in `NullReferenceException`'s due to the data race.

### Fix

I used a few `DelegateBasedAsyncTask`'s to fix it, but I'm not yet really happy with the solution.  I have to keep the synchronous method `RefreshExpandedDisplay` around as making the switch to async causes bizarre issues with `BeginExpandTransition`.  The preview seems to immediately be hidden, even if I schedule the transition to take place after changes in the UI.  

### Reviewed

- [x] @Benglin

### CODE REVIEW

File: src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs [hunk 1]
```diff
@@ -264,6 +271,52 @@ private void RefreshCondensedDisplay()
...
+            var watchTree = largeContentGrid.Children[0] as WatchTree;
+            var rootDataContext = watchTree.DataContext as WatchViewModel;
+
+            var task0 = new DelegateBasedAsyncTask(scheduler, () =>
+            {
+                cachedLargeContent = nodeViewModel.DynamoViewModel.WatchHandler.GenerateWatchViewModelForData(
+                    mirrorData, null, string.Empty, false);
+            });
+
+            scheduler.ScheduleForExecution(task0);
```
Comment (Benglin) [inline src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs]: We don't need the second task below as we can count on the completion notification from the first task (do note that the callback for `Completed` event is done on scheduler thread, which is why we still need a `BeginInvoke` call):

```csharp
task0.Completed += delegate(AsyncTask task)
{
    Dispatcher.BeginInvoke((Action) (() =>
    {
        rootDataContext.Children.Add(cachedLargeContent);

        watchTree.treeView1.SetBinding(ItemsControl.ItemsSourceProperty,
            new Binding("Children")
            {
                DynamoModel.MakeConnectionCommand.Mode = BindingMode.TwoWay,
                Source = rootDataContext
            });

        BeginViewSizeTransition(ComputeLargeContentSize());
    }));
};

scheduler.ScheduleForExecution(task0);
```

Please take note that I have moved `BeginViewSizeTransition` from the caller.

File: src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs [hunk 2]
```diff
@@ -135,8 +139,11 @@ internal void BindToDataSource(MirrorData mirrorData)
             }
             else if (this.IsExpanded)
             {
-                RefreshExpandedDisplay();
-                BeginViewSizeTransition(ComputeLargeContentSize());
+                RefreshExpandedDisplayAsync();
+
+                scheduler.ScheduleForExecution(
+                    new DelegateBasedAsyncTask(scheduler, 
+                        () => Dispatcher.BeginInvoke((Action)(() => BeginViewSizeTransition(ComputeLargeContentSize())))));
```
Comment (Benglin) [inline src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs]: This call to `BeginViewSizeTransition` is to be moved into `RefreshExpandedDisplayAsync` (see below for details).

File: src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs [hunk 3]
```diff
@@ -264,6 +271,52 @@ private void RefreshCondensedDisplay()
...
+                        {
+                            Mode = BindingMode.TwoWay,
+                            Source = rootDataContext
+                        });
+                })));
+
+            scheduler.ScheduleForExecution(task1);
+        }
+
         private void RefreshExpandedDisplay()
```
Comment (Benglin) [inline src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs]: What do we do for this one? Seems like it's also a dangerous method which may cause race condition, do we get rid of it?

File: src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs [hunk 4]
```diff
@@ -264,6 +271,52 @@ private void RefreshCondensedDisplay()
...
+                largeContentGrid.Children.Add(tree);
+            }
+
+            var watchTree = largeContentGrid.Children[0] as WatchTree;
+            var rootDataContext = watchTree.DataContext as WatchViewModel;
+
+            var task0 = new DelegateBasedAsyncTask(scheduler, () =>
+            {
+                cachedLargeContent = nodeViewModel.DynamoViewModel.WatchHandler.GenerateWatchViewModelForData(
+                    mirrorData, null, string.Empty, false);
```
Comment (Benglin) [inline src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs]: I tried to ignore this but couldn't, `cachedLargeContent` is being set in the context of `SchedulerThread`, but it is also being referred to by other parts of the class (which run on UI thread). I think the safest way to deal with this is to have a local version of `var largeContent`, then use it later on to update `cachedLargeContent` during `Dispatcher.BeginInvoke`. What are your thoughts on this?

**Comment by Benglin** [comment]
Thanks for the fix, @pboyer. I think I can deal with getting rid of synchronous version of `RefreshExpandedDisplay`, can you make a *Critical* task for me to get rid of that one after this crash is fixed?

File: src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs (hunk 4 above)
Comment (pboyer) [inline src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs]: My intuition was to do this actually!

File: src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs (hunk 2 above)
Comment (pboyer) [inline src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs]: I generalized this a bit.  The caller can pass in an `Action` that can be executed once `RefreshExpandedDisplay` but on the UI thread.

File: src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs (hunk 3 above)
Comment (pboyer) [inline src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs]: I would very much like to be rid of it, but as discussed above, I could not figure out how to remove it.

**Comment by pboyer** [comment]
Hi Ben,

I added a few small small things.  Namely, I added a few extension methods that make it easier to compose actions to perform after a task is completed.  These are in `AsyncExtensions`.  

I've also created a task to remove the synchronous version of `RefreshExpandedDisplay.`  This is likely to cause race conditions once the timer work is in: 

http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-6444

**Comment by Benglin** [comment]
Brilliant! Thanks @pboyer, I have learnt a new thing today ;)

**Pull request merged**